doc/build-overview.md

```diff
@@ -0,0 +1,251 @@
+  project package, and add that package + component to the set of
+  project targets.
+* Ensure that no target has been specified multiple times.
+
+We now have an update four package maps, a new set of dependency
+targets, and a new set of project package targets (potentially with
+specific components).
+
+## Apply named CLI flags
+
+Named CLI flags are applied to specific packages by updating the
+config in one of the four maps. If a flag is specified and no package
+is found, it's an error.
+
+## Apply CLI GHC options
+
+Apply GHC options from the command line to all _project package
+targets_. *FIXME* confirm that this is in fact the correct behavior.
+
+## Apply general flags (CLI and config value)
+
+*FIXME* figure out and document exactly which packages these will
+apply to.
+
+## Apply general GHC options
+
+*FIXME* list out the various choices here and which packages they
+apply to.
+
+## Determine snapshot hash
+
+Use some deterministic binary serialization and SHA256 thereof to get
+a hash of the following information:
+
+* Actual compiler (GHC version, path, *FIXME* probably some other
+  unique info from GHC, I've heard that `ghc --info` gives you
+  something)
+* Global database map
+* Immutable dependency map
+
+Motivation: Any package built from the immutable dependency map and
+installed in this database will never need to be rebuilt.
+
+*FIXME* Caveat: do we need to take profiling settings into account
+here? How about Haddock status?
+
+## Determine actual target components
+
+* Dependencies: "default" components (all libraries and executables)
+* Project packages:
+    * If specific components named: only those, plus any libraries present
+    * If no specific components, include the following:
+        * All libraries, always
+        * All executables, always
+        * All test suites, _if_ `--test` specified on command line
+        * All benchmarks, _if_ `--bench` specified on command line
+
+## Construct build plan
+
+* Applied to every target (project package or dependency)
+* Apply flags, platform, and actual GHC version to resolve
+  dependencies in any package analyzed
+* Include all library dependencies for all enabled components
+* Include all build tool dependencies for all enabled components
+  (using the fun backwards compat logic for `build-tools`)
+* Apply the logic recursively to come up with a full build plan
+* If a task depends exclusively on immutable packages, mark it as
+  immutable. Otherwise, it's mutable. The former go into the snapshot
+  database, the latter into the local database.
+
+We now have a set of tasks of packages/components to build, with full
+config information for each package, and dependencies that must be
+built first.
+
+*FIXME* There's some logic to deal with cyclic dependencies between
+test suites and benchmarks, where a task can be broken up into
+individual components versus be kept as a single task. Need to
+document this better. Currently it's the "all in one" logic.
```

Comment:
Backpack would hook into this step: when you were configuring a component, you'd also be obligated to mix-in link it (assuming that the things that it referred to have recursively already been resolved).

Then, after everything is said and done, you can go over *fully instantiated* references to packages, and add extra jobs to the build graph to build the instantiated forms of those packages.

Reply:
You're still able to create a complete build graph during plan construction. Actually, I suppose you can add the instantiated packages as you're initially building the build graph, but I think it's easier to do split it into two phases, just from a code complexity perspective.